libkbfs/quota_usage.go

```diff
@@ -0,0 +1,134 @@
+// NewEventuallyConsistentQuotaUsage creates a new
+// EventuallyConsistentQuotaUsage object.
+func NewEventuallyConsistentQuotaUsage(
+	config Config) *EventuallyConsistentQuotaUsage {
+	return &EventuallyConsistentQuotaUsage{
+		config: config,
+		log:    config.MakeLogger(ECQUID),
+	}
+}
+
+func (q *EventuallyConsistentQuotaUsage) getAndCache(
+	ctx context.Context) (usage cachedQuotaUsage, err error) {
+	defer func() {
+		q.log.CDebugf(ctx, "getAndCache: error=%v", err)
+	}()
+	quotaInfo, err := q.config.BlockServer().GetUserQuotaInfo(ctx)
+	if err != nil {
+		return usage, err
+	}
+
+	usage.limitBytes = quotaInfo.Limit
+	if quotaInfo.Total != nil {
+		usage.usageBytes = quotaInfo.Total.Bytes[kbfsblock.UsageWrite]
+	} else {
+		usage.usageBytes = 0
+	}
+	usage.timestamp = time.Now()
+
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	q.cached = usage
+
+	return usage, nil
+}
+
+// Get returns KBFS bytes used and limit for user. To help avoid having too
+// frequent calls into bserver, caller can provide a positive tolerance, to
+// accept stale LimitBytes and UsageByts data.
+//
+// 1) If the cached (stale) data is older for less than half of tolerance, the
+// stale data is returned immediately.
+// 2) If the cached (stale) data is older for more than half of tolerance, but
+// not more than tolerance, a background RPC is spawned to refresh cached data,
+// and the stale data is returned immediately.
+// 3) If the cached (stale) data is older for more than tolerance, a blocking
+// RPC is issued and the function only returns after RPC finishes, with the
+// newest data from RPC.
+func (q *EventuallyConsistentQuotaUsage) Get(ctx context.Context,
+	tolerance time.Duration) (usageBytes, limitBytes int64, err error) {
+	c := func() cachedQuotaUsage {
+		q.mu.RLock()
+		defer q.mu.RUnlock()
+		return q.cached
+	}()
+	past := time.Since(c.timestamp)
+	switch {
+	case past > tolerance:
+		q.log.CDebugf(ctx, "Blocking on getAndCache. Cached data is %s old.", past)
+		c, err = q.getAndCache(ctx)
+		if err != nil {
+			return -1, -1, err
+		}
+	case past > tolerance/2:
+		if atomic.CompareAndSwapInt32(&q.backgroundInProcess, 0, 1) {
+			id, err := MakeRandomRequestID()
+			if err != nil {
+				q.log.Warning("Couldn't generate a random request ID: %v", err)
+			}
+			q.log.CDebugf(ctx, "Cached data is %s old. Spawning getAndCache in "+
+				"background with tag:%s=%v.", past, ECQUID, id)
+			go func() {
+				// Make a new context so that it doesn't get canceled when returned.
+				logTags := make(logger.CtxLogTags)
+				logTags[ECQUCtxTagKey{}] = ECQUID
+				bgCtx := logger.NewContextWithLogTags(context.Background(), logTags)
+				bgCtx = context.WithValue(bgCtx, ECQUCtxTagKey{}, id)
+				// Make sure a timeout is on the context, in case the RPC blocks
+				// forever somehow, where we'd end up with never resetting
+				// backgroundInProcess flag again.
+				bgCtx, cancel := context.WithTimeout(bgCtx, 10*time.Second)
+				defer cancel()
+				q.getAndCache(bgCtx)
```

Comment:
`getAndCache()` already logs the error in a deferred block. Is there any added information to log here?